### PR TITLE
Add text_generation_util to support TopK, TopP sampling + Tabular Data Generation.

### PR DESCRIPTION
**Pull request by yidong72**
# What does this PR do?

Added the text_generation_util to support the MegatronLM way of text generation. It supports top_k, top_p sampling. 

Added the Tabular Data tokenizer which recognizes the tabular data structure. 
Added the unit test for column type encode and decode.
Added tabular data generation which considers the tabular data structure.

To run text generation. here is one example.
```python
python examples/nlp/language_modeling/megatron_gpt_eval.py \
            model_file=/path/gpt_model.nemo
```
the inference parameters are set in file
```
megatron_gpt_inference.yaml
```
Update: 

Included GPT inference server. To use it  
```python
python examples/nlp/language_modeling/megatron_gpt_eval.py \
            model_file=/path/gpt_model.nemo \
            server=True \
            port=5555
```
Here is the example code to run inference as a client
```python
import json
import requests

batch_size = 8
port_num = 5555
headers = {"Content-Type": "application/json"}


def request_data(data):
    resp = requests.put('http://localhost:{}/generate'.format(port_num),
                        data=json.dumps(data),
                        headers=headers)
    sentences = resp.json()['sentences']
    return sentences


data = {
    "sentences": [""] * batch_size,
    "tokens_to_generate": 300,
    "temperature": 1.0,
    "add_BOS": True,
    "top_k": 0,
    "top_p": 0.9,
    "greedy": False,
    "all_probs": False,
    "repetition_penalty": 1.2,
    "min_tokens_to_generate": 2,
}

sentences = request_data(data)

for i in range(batch_size):
    s = sentences[i]
    print(s)
```

### CODE REVIEW

**Review by ericharper — CHANGES_REQUESTED** [review]
We should not have two eval scripts: can you modify `megatron_gpt_eval.py` ?

**Comment by yidong72** [comment]
I tested the Pipeline model parallelism GPT, and the inference works.

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 1]
```diff
@@ -12,240 +12,101 @@
...
+        plugins=NLPDDPPlugin(),
+        devices=cfg.tensor_model_parallel_size * cfg.pipeline_model_parallel_size,
         accelerator='gpu',
-        precision=args.precision,
+        precision=cfg.precision,
     )
 
-    if args.model_file:
-        model = MegatronGPTModel.restore_from(restore_path=args.model_file, trainer=trainer)
-    elif args.checkpoint_dir:
```
Comment (ericharper) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:169]: Why remove this?

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  devices: 1
+  num_nodes: 1
+  accelerator: gpu
+  logger: False # logger provided by exp_manager
+  precision: 16 # 16 or 32
+
+tensor_model_parallel_size: 1
+pipeline_model_parallel_size: 1
+model_file: null  # GPT nemo file path
+checkpoint_dir: null # checkpoint dir
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:20]: can you specify this PTL checkpoint from training

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+checkpoint_name: null # checkpoint name, only used for checkpoint loading
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:21]: add PTL to comment

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+hparams_file: null # model configuration file, only used for checkpoint loading
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:22]: Add PTL to comment

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 5]
```diff
@@ -12,240 +12,101 @@
...
+    if cfg.precision in ["32", "16"]:
+        cfg.precision = int(float(cfg.precision))
 
     # trainer required for restoring model parallel models
     trainer = Trainer(
-        plugins=[NLPDDPPlugin()],
-        devices=args.devices,
-        num_nodes=args.num_nodes,
+        plugins=NLPDDPPlugin(),
+        devices=cfg.tensor_model_parallel_size * cfg.pipeline_model_parallel_size,
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_gpt_eval.py]: this is fixed

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 6]
```diff
@@ -12,240 +12,101 @@
...
+class RequestDataSet(Dataset):
+    def __init__(self, sentences):
+        super().__init__()
+        self.sentences = sentences
 
-    args = parser.parse_args()
+    def __len__(self,):
+        return len(self.sentences)
 
-    assert (
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:147]: yes. I will add it

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml (hunk 2 above)
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml]: done

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml (hunk 3 above)
Comment (yidong72) [reply examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml]: done

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+    return tokens, attention_mask, position_ids
+
+
+def tab_logits(logits, min_id, max_id, filter_value=-float('Inf')):
+    logits[:, :min_id] = filter_value
+    logits[:, max_id:] = filter_value
+    return logits
+
+
+def top_k_logits(logits, top_k=0, top_p=0.0, filter_value=-float('Inf')):
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:59]: Very useful!

File: scripts/tokenizers/conf/tabular_data_tokenizer.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+      hasnan: True
+  - name: col_d
+    code_type: category
+    args:
+      code_len: 3
+      base: 12
+      fillall: True
+      hasnan: True
+tokenizer_file: ???  # tabular tokneizer output file path
+table_csv_file: ???  # input table csv file
```
Comment (michalivne) [inline scripts/tokenizers/conf/tabular_data_tokenizer.yaml:33]: End with newline?

**Review by michalivne — DISMISSED** [review]
Fantastic PR, thanks! LGTM!
See very minor comments

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 9]
```diff
@@ -12,166 +12,60 @@
...
-            --prompt=YOUR_PROMPT
-"""
+try:
+    from apex.transformer import parallel_state
+
+    HAVE_APEX = True
+except (ImportError, ModuleNotFoundError):
+    HAVE_APEX = False
 
 assert torch.cuda.is_available()
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_gpt_eval.py]: fixed

File: nemo/collections/common/tokenizers/column_coder.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+        return [self.item_to_id[item]]
+
+    def decode(self, ids: List[int]) -> str:
+        return self.id_to_item[ids[0]]
+
+
+column_map = {"int": IntCode, "float": FloatCode, "category": CategoryCode}
+
+
+class ColumnCodes:
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/common/tokenizers/column_coder.py]: fixed

File: scripts/tokenizers/conf/tabular_data_tokenizer.yaml (hunk 8 above)
Comment (yidong72) [reply scripts/tokenizers/conf/tabular_data_tokenizer.yaml]: fixed

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 11]
```diff
@@ -194,6 +200,15 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
         else:
             raise ValueError('precision must be in [32, 16, "bf16"]')
 
+        # configuration used for inference
+        self.__inference_config = None
+
+    def set_inference_config(self, inference_config):
+        self.__inference_config = inference_config
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: changed

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 12]
```diff
@@ -829,193 +878,264 @@ def prompt_tuning_param_freeze_and_optimizer_setup(self):
...
         for param_group in self._optimizer_param_groups:
             for param in param_group['params']:
                 params.append(param)
         return params
 
-    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
-        request, positions, tokens_to_generate, compute_logprobs = MegatronGPTModel._bucketize_gpt_inference(
-            batch, self.use_soft_prompts
-        )
+    def __get_computeprob_response(self, response, inputs):
```
[the remark this answers is not in the document]
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: changed

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 13]
```diff
@@ -12,166 +12,62 @@
...
-    )
-    parser.add_argument(
-        "--checkpoint_name",
-        type=str,
-        default=None,
-        required=False,
-        help="If not using a .nemo file. Name of checkpoint to be used. Ex: megatron_gpt--val_loss=6.34-step=649-last.ckpt",
-    )
+try:
+    from apex.transformer import parallel_state
```
Comment (michalivne) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:33]: Can we add back a docstring with usage examples?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 14]
```diff
@@ -829,193 +878,264 @@ def prompt_tuning_param_freeze_and_optimizer_setup(self):
...
+        self,
+        inputs: Union[List[str], torch.Tensor, List[dict]],
+        length_params: LengthParam,
+        sampling_params: SamplingParam = None,
+    ) -> OutputType:
+
+        # check whether the DDP is initialized
+        if parallel_state.is_unitialized():
+
+            class RequestDataSet(Dataset):
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:930]: Use here the shared class from `dataset_utils`?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 15]
```diff
@@ -829,193 +878,264 @@ def prompt_tuning_param_freeze_and_optimizer_setup(self):
...
         for param_group in self._optimizer_param_groups:
             for param in param_group['params']:
                 params.append(param)
         return params
 
-    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
-        request, positions, tokens_to_generate, compute_logprobs = MegatronGPTModel._bucketize_gpt_inference(
-            batch, self.use_soft_prompts
-        )
+    def _get_computeprob_response(self, response, inputs):
```
Comment (michalivne) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:888]: Can we make this method shared as well?
The same code will be used by all Megatron NLP models, right?

File: nemo/collections/nlp/modules/common/text_generation_utils.py [hunk 16]
```diff
@@ -0,0 +1,773 @@
...
+                output = output[0]['logits'].float()
+                assert output is not None
+                output = output.float()
+                logits = output[:, -1].view(batch_size, -1).contiguous()
+
+                # make sure it will generate at least min_length
+                min_length = extra.get('min_tokens_to_generate', 0)
+                if min_length > 0:
+                    within_min_length = (context_length - context_lengths) < min_length
+                    logits[within_min_length, eod_id] = -float('Inf')
```
Comment (michalivne) [inline nemo/collections/nlp/modules/common/text_generation_utils.py:514]: Thanks for adding the filtering of padded tokens.

**Review by michalivne — COMMENTED** [review]
Great PR! See minor comments and suggestions.

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 17]
```diff
@@ -12,166 +12,62 @@
...
-        required=False,
-        help="True/False: whether to stop after full sentence has been generated.",
-    )
-    parser.add_argument("--batch_size", default=1, type=int, required=False, help="Evaluation batch_size")
-    parser.add_argument(
-        "--compute_logprobs", type=bool, default=False, required=False, help="Method for logprobs computation"
-    )
 
-    args = parser.parse_args()
+class RequestDataSet(Dataset):
```
[the remark this answers is not in the document]
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:43]: This is a dummy dataset used only for trainer.predict text generation method. I am not sure anyone will use it outside of gpt eval.

File: examples/nlp/language_modeling/megatron_gpt_eval.py (hunk 13 above)
Comment (yidong72) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:33]: Good point. will add a docstring.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 14 above)
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:930]: this is just a hack to invoke the ddp initialization. that's why I use it as an inner class. Need to fix this in the future when we can manually do ddp initialization.

**Pull request merged**